### PR TITLE
Manage solidity types

### PR DESCRIPTION
**Pull request by yann300**
Manage solidity types:
	- uint<N>, uint
	- int<N>, int
	- hash<N>, hash, address
	- bool
	- string<N>

### CODE REVIEW

File: libdevcore/CommonJS.cpp [hunk 1]
```diff
@@ -45,6 +45,13 @@ bytes padded(bytes _b, unsigned _l)
 	return asBytes(asString(_b).substr(_b.size() - std::max(_l, _l)));
 }
 
+bytes paddedRight(bytes _b, unsigned _l)
+{
+	while (_b.size() < _l)
```
Comment (arkpar) [inline libdevcore/CommonJS.cpp]: Why not just _b.resize(_l); return _b; ?
Is it supposed to append zero bytes or '0' - zero character? It is not clear from the documentation comment.

File: mix/QVariableDefinition.cpp [hunk 2]
```diff
@@ -53,3 +55,95 @@ QVariableDefinition* QVariableDefinitionList::val(int _idx)
...
+dev::bytes QStringType::encodeValue()
+{
+	qDebug() << QString::fromStdString(toJS(paddedRight(asBytes(value().toStdString()), 32)));
+	return paddedRight(asBytes(value().toStdString()), 32);
+}
+
+void QStringType::decodeValue(std::string const& _rawValue)
+{
+	std::string rawParam = _rawValue.substr(0, 32 * 2);
+	rawParam = unpadRight(rawParam);
```
Comment (arkpar) [inline mix/QVariableDefinition.cpp]: Why would you remove trailing zero characters?

File: libdevcore/CommonJS.cpp [hunk 3]
```diff
@@ -52,6 +59,18 @@ bytes unpadded(bytes _b)
 	return _b;
 }
 
+std::string unpadRight(std::string _b)
```
Comment (arkpar) [inline libdevcore/CommonJS.cpp]: Can you give a case for this function to be required?

File: mix/QVariableDefinition.cpp (hunk 2 above)
Comment (yann300) [inline mix/QVariableDefinition.cpp]: From the Wiki, we need to right aligned the string: 
string<N>: binary type of N bytes, N > 0. Assumed to be UTF-8 encoded, right-aligned, zero-passed data.

File: mix/ClientModel.cpp [hunk 4]
```diff
@@ -168,15 +171,8 @@ void ClientModel::executeSequence(std::vector<TransactionSettings> const& _seque
...
-				{
-					QVariableDeclaration* var = (QVariableDeclaration*)f->parametersList().at(p);
-					u256 value = 0;
-					auto v = t.parameterValues.find(var->name());
-					if (v != t.parameterValues.cend())
-						value = v->second;
-					c.encode(var, value);
-				}
+				for (int p = 0; p < t.parameterValues.size(); p++)
+					c.push(t.parameterValues.at(p)->encodeValue());
```
Comment (arkpar) [inline mix/ClientModel.cpp]: There needs to be a check if stored parameter type matches actual parameter type for the contract. The contract could have been edited after the state was configured

File: mix/ContractCallDataEncoder.cpp [hunk 5]
```diff
@@ -74,59 +57,21 @@ QList<QVariableDefinition*> ContractCallDataEncoder::decode(QList<QVariableDecla
...
+		if (dec->type().contains("int"))
+			def = new QIntType(dec, QString());
+		else if (dec->type().contains("real"))
+			def = new QRealType(dec, QString());
+		else if (dec->type().contains("bool"))
+			def = new QBoolType(dec, QString());
+		else if (dec->type().contains("string") || dec->type().contains("text"))
+			def = new QStringType(dec, QString());
+		else if (dec->type().contains("hash") || dec->type().contains("address"))
+			def = new QHashType(dec, QString());
```
Comment (arkpar) [inline mix/ContractCallDataEncoder.cpp]: else throw exception probably, otherwise it is a segfault

File: libdevcore/CommonJS.cpp (hunk 1 above)
Comment (yann300) [inline libdevcore/CommonJS.cpp]: This is supposed to be zero bytes. so we can use _b.resize(_l)

File: mix/QVariableDeclaration.h [hunk 6]
```diff
@@ -38,7 +40,8 @@ class QVariableDeclaration: public QBasicNodeDefinition
 	QVariableDeclaration() {}
 	QVariableDeclaration(solidity::VariableDeclaration const* _v): QBasicNodeDefinition(_v), m_type(QString::fromStdString(_v->getType()->toString())) {}
 	QVariableDeclaration(solidity::ParamDescription const& _v): QBasicNodeDefinition(_v.getName()), m_type(QString::fromStdString(_v.getType())) {}
-	QString type() const { return m_type; }
+	Q_INVOKABLE QString type() const { return m_type; }
```
Comment (arkpar) [inline mix/QVariableDeclaration.h]: Better make if a property. Q_PROPERTY(QString type MEMBER m_type CONSTANT)

File: mix/ContractCallDataEncoder.cpp [hunk 7]
```diff
@@ -44,29 +44,12 @@ void ContractCallDataEncoder::encode(QFunctionDefinition const* _function)
...
-	bytes data = padded(jsToBytes(s.str()), padding);
-	m_encodedData.insert(m_encodedData.end(), data.begin(), data.end());
-	encodedData();
-}
-
-QList<QVariableDefinition*> ContractCallDataEncoder::decode(QList<QVariableDeclaration*> _returnParameters, bytes _value)
+QList<QVariableDefinition*> ContractCallDataEncoder::decode(QList<QVariableDeclaration*> const& _returnParameters, bytes const& _value)
 {
 	QList<QVariableDefinition*> r;
 	std::string returnValue = toJS(_value);
```
Comment (arkpar) [inline mix/ContractCallDataEncoder.cpp]: I think this conversion to string is not needed anymore

File: mix/QVariableDefinition.cpp [hunk 8]
```diff
@@ -53,3 +55,95 @@ QVariableDefinition* QVariableDefinitionList::val(int _idx)
...
+{
+	dev::bigint i(value().toStdString());
+	if (i < 0)
+		i = i + dev::bigint("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff") + 1;
+	std::ostringstream s;
+	s << std::hex << "0x" << i;
+	return padded(jsToBytes(s.str()), 32);
+}
+
+void QIntType::decodeValue(std::string const& _rawValue)
```
Comment (arkpar) [inline mix/QVariableDefinition.cpp]: Why encode to bytes directly but decode to string and then from string to value?

File: mix/ClientModel.cpp (hunk 4 above)
Comment (yann300) [inline mix/ClientModel.cpp]: ok so I guess that check should be earlier in the use case.
Like : 
When the user click on "Run State" (before entering in the debugState function).
When the user edit a state (we can display transaction that are not in scope in red).

File: libdevcore/CommonJS.cpp (hunk 3 above)
Comment (yann300) [inline libdevcore/CommonJS.cpp]: This has to be used only for real string. But the input parameter is a string not an int ...
Probably should add comment to warn user to not use with a non parse int.

File: mix/QVariableDefinition.cpp [hunk 9]
```diff
@@ -53,3 +55,95 @@ QVariableDefinition* QVariableDefinitionList::val(int _idx)
 		return nullptr;
 	return m_def.at(_idx);
 }
+
+/*
+ * QIntType
+ */
+dev::bytes QIntType::encodeValue()
+{
+	dev::bigint i(value().toStdString());
```
Comment (arkpar) [inline mix/QVariableDefinition.cpp]: I think you can just use dev::toBigEndian here

File: mix/QVariableDefinition.cpp (hunk 8 above)
Comment (arkpar) [inline mix/QVariableDefinition.cpp]: Consider using dev::fromBigEndian

Comment (yann300) [inline mix/QVariableDefinition.cpp]: so we need to encode to byte in order to call m_client.transact(). And we need to decode to string ... to use the return parameter. I don't get your point here.

File: mix/QVariableDefinition.cpp [hunk 10]
```diff
@@ -53,3 +55,95 @@ QVariableDefinition* QVariableDefinitionList::val(int _idx)
...
+/*
+ * QStringType
+ */
+dev::bytes QStringType::encodeValue()
+{
+	qDebug() << QString::fromStdString(toJS(paddedRight(asBytes(value().toStdString()), 32)));
+	return paddedRight(asBytes(value().toStdString()), 32);
+}
+
+void QStringType::decodeValue(std::string const& _rawValue)
```
Comment (arkpar) [inline mix/QVariableDefinition.cpp]: This should not require converting a string to hex string and back

File: mix/ContractCallDataEncoder.cpp (hunk 7 above)
Comment (yann300) [inline mix/ContractCallDataEncoder.cpp]: At some point we need to convert to string to extract the return parameters ...

File: mix/QVariableDefinition.cpp (hunk 10 above)
Comment (yann300) [inline mix/QVariableDefinition.cpp]: for a string, the return value is hex encoded, so we need to convert to ASCII char.

File: mix/ClientModel.cpp (hunk 4 above)
Comment (arkpar) [inline mix/ClientModel.cpp]: Checking on run will be enough for now I suppose, This should be done in this method

File: libdevcore/CommonJS.cpp [hunk 11]
```diff
@@ -52,12 +58,21 @@ bytes unpadded(bytes _b)
...
 {
-	auto p = _b.find_first_not_of('0');
-	if (p == std::string::npos)
-		return "0";
-	return _b.substr(p, _b.length() - 1);
+	int i = 0;
+	if (_b.size() == 0)
+		return _b;
+	while (true)
+	{
```
Comment (gavofyork) [inline libdevcore/CommonJS.cpp]: no need for {s

File: libdevcore/CommonJS.h [hunk 12]
```diff
@@ -48,12 +48,14 @@ inline std::string toJS(dev::bytes const& _n)
...
 bytes jsToBytes(std::string const& _s);
-/// Add '0' on the head of _b until _l.
+/// Add '0' on the head of @a _b until @a _l.
 bytes padded(bytes _b, unsigned _l);
+/// Add '0' on the queue of @a _b until @a _l.
+bytes paddedRight(bytes _b, unsigned _l);
 /// Removing all trailing '0'. Returns empty array if input contains only '0' char.
 bytes unpadded(bytes _s);
-/// Remove all '0' on the head of _s. Returns 0 if _s contains only '0'.
-std::string unpadLeft(std::string _s);
```
Comment (gavofyork) [inline libdevcore/CommonJS.h]: this alteration of a base api function doesn't break stuff in the rest of the project?

Comment (yann300) [inline libdevcore/CommonJS.h]: This function was only used in mix and it is not used anymore.

File: libdevcore/CommonJS.cpp [hunk 13]
```diff
@@ -52,12 +58,21 @@ bytes unpadded(bytes _b)
...
 {
-	auto p = _b.find_first_not_of('0');
-	if (p == std::string::npos)
-		return "0";
-	return _b.substr(p, _b.length() - 1);
+	int i = 0;
+	if (_b.size() == 0)
+		return _b;
+
+	while (true)
```
Comment (arkpar) [inline libdevcore/CommonJS.cpp]: Will cause exception is if there is no zero byte in the vector.
You can use 
while(i < _b.size() && _b[i] == 0)
   i++;

File: mix/ClientModel.cpp [hunk 14]
```diff
@@ -204,16 +208,14 @@ void ClientModel::executeSequence(std::vector<TransactionSettings> const& _seque
...
-						auto v = transaction.parameterValues.find(var->name());
-						if (v != transaction.parameterValues.cend())
-							value = v->second;
-						encoder.encode(var, value);
+						if (f->parametersList().at(p)->type() != transaction.parameterValues.at(p)->declaration()->type())
+							BOOST_THROW_EXCEPTION(ParameterChangedException() << FunctionName(f->parametersList().at(p)->type().toStdString()));
+						encoder.push(transaction.parameterValues.at(p)->encodeValue());
 					}
 
+
```
Comment (arkpar) [inline mix/ClientModel.cpp]: extra linebreak

File: mix/ClientModel.cpp [hunk 15]
```diff
@@ -224,7 +226,9 @@ void ClientModel::executeSequence(std::vector<TransactionSettings> const& _seque
 						}
 					}
 					else
+					{
```
Comment (arkpar) [inline mix/ClientModel.cpp]: extra braces

File: mix/QVariableDefinition.cpp [hunk 16]
```diff
@@ -53,3 +55,92 @@ QVariableDefinition* QVariableDefinitionList::val(int _idx)
...
+{
+	m_bigIntvalue = _value;
+	std::stringstream str;
+	str << std::dec << m_bigIntvalue;
+	m_value = QString::fromStdString(str.str());
+}
+
+dev::bytes QIntType::encodeValue()
+{
+	dev::bigint i(value().toStdString());
```
Comment (arkpar) [inline mix/QVariableDefinition.cpp]: can't m_bigIntvalue be used here?

File: mix/QVariableDefinition.cpp [hunk 17]
```diff
@@ -53,3 +55,92 @@ QVariableDefinition* QVariableDefinitionList::val(int _idx)
...
+		bigint = bigint - dev::bigint("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff") - 1;
+	setValue(bigint);
+}
+
+/*
+ * QHashType
+ */
+dev::bytes QHashType::encodeValue()
+{
+	QByteArray b = value().toUtf8();
```
Comment (arkpar) [inline mix/QVariableDefinition.cpp]: I don't think you need toUtf8 here for hash

File: mix/QVariableDefinition.cpp [hunk 18]
```diff
@@ -53,3 +55,92 @@ QVariableDefinition* QVariableDefinitionList::val(int _idx)
...
+{
+	setValue(QString::fromUtf8((char*)_rawValue.data()));
+}
+
+/*
+ * QBoolType
+ */
+dev::bytes QBoolType::encodeValue()
+{
+	return padded(jsToBytes(value().toStdString()), 32);
```
Comment (arkpar) [inline mix/QVariableDefinition.cpp]: can't m_boolValue be used here?

File: mix/QVariableDefinition.cpp (hunk 16 above)
Comment (yann300) [inline mix/QVariableDefinition.cpp]: At this point, QIntType comes from QML, so we only have the string value. 
Maybe we can add a function to resolve the bigint value in QML side for thoses specific types (bool, int, uint).

Comment (yann300) [inline mix/QVariableDefinition.cpp]: but does it worth it?

File: libdevcore/CommonJS.cpp (hunk 13 above)
Comment (arkpar) [inline libdevcore/CommonJS.cpp]: My mistake, it is ok

File: mix/QVariableDefinition.cpp (hunk 18 above)
Comment (yann300) [inline mix/QVariableDefinition.cpp]: Same as for m_bigIntValue.

**Comment by yann300** [comment]
When adding a function containing parameters, String<N> types should be managed by mix (as well as uint<N>, int<N>, hash<N>, bool). Return parameters are shown in the transaction logs table view (column returned).

**Pull request merged**